### ISSUE TITLE
Workaround: simply add support for vulkan and opengl support

### ISSUE THREAD

**Issue by simonlsp** [jamesstringerparsec/Easy-GPU-PV#342]
## Problem:
* Vulkan and opengl are not supported while streaming.

## Cause:
* Video is streamed from parsec virtual monitor, idd monitor or usbmidd monitor.
* The hyper-v monitor (which is logicially connected to the hyper-v gpu and PV gpu) is off while streaming.
* Without connected monitor, serveal PV gpu functions were disableed, including vulkan and opengl.

## Solution:
* Enable hyper-V gpu in device manager.
* Open display setting while streaming.
* Click "detect" in multi monitor menu.
* Hyper-v monitor should be detected (without need for host connection).
* GPU-PV vulkan and opengl will be available for **both** monitor now.

## Known problems and workarounds
* Vulkan and opengl will only work if Hyper-V monitor is on, if you set the display mod to "only display on the other monitor", vulkan and opengl will be disabled again. 
  * Solution: set your display mode to "extend", and arrange monior layout like this, so your mouse pointer would not fly away.
![image](https://github.com/jamesstringerparsec/Easy-GPU-PV/assets/23612863/8c87a934-c1fc-4203-87a3-0b7a0c7eb9b8)
* Hyper-V monitor will always be 1st monitor, however ,we want to stream from virtual monitor for better performance.
  *  Solution: use [autohotkey](https://www.autohotkey.com/) and MultiMonitorTool to create a hotkey to swith primary monitor, and use sunshine for streaming. Sunshine use primary monitor for streaming if no streaming monitor is set.

## Result 

All following results were captured while sunshine streaming.

![image](https://github.com/jamesstringerparsec/Easy-GPU-PV/assets/23612863/c4a9adfb-6d21-42cc-ace6-84f2e6a2db9e)
![image](https://github.com/jamesstringerparsec/Easy-GPU-PV/assets/23612863/082e6e1f-a06d-4c7f-9bb7-3b1e2e61a713)

**Comment by simonlsp** [comment]
@Kodikuu 
With this workaround, I think it is time to change this line in the readme notes.
> Vulkan renderer is unavailable and GL games may or may not work.
Yuzu emu and minecraft were tested, they worked flawlessly.

**Comment by KenZync** [comment]
I can't run blender with this solution

**Comment by ichinoboy** [comment]
Hi @simonlsp ,

I tried this in Windows 11 and it won't allow me to extend my monitor to 2nd one (Hyper-V) as it is not active. Do you have a workaround for this?

**Comment by iso2013** [comment]
When I change the display settings on my instance of Parsec, the parsec connection hangs (completely unresponsive) and I have to restart my PC (killing Parsec from task manager freezes the rest of the PC. Not really sure why)

**Comment by djblue** [comment]
In trying out the above, I ran into some issues but found what seems to be an optimal solution for myself. I have no idea why this is working for me, but here are the tweaked steps that seem to work:

- Connect via Parsec / Moonlight
- Install the extra display with [IddSampleDriver](https://github.com/ge9/IddSampleDriver) via scoop.
  - Not sure if this is only needed because I'm using Sunshine
- Install [FurMark](https://www.geeks3d.com/furmark/) to help detect if OpenGL and Vulkan are available
  - Running it at this point should fail
 - Re-enable the `Microsoft Hyper-V Video` Display adapter in Device manager
- Connect to the vm display via Hyper-V, while still connected via (Parsec / Moonlight)
  - Windows should now register two displays
- FurMark should now be able to open and you can confirm OpenGL is working
  - Vulkan doesn't work at this point for me
- Close the display for Hyper-V
- Now windows should only  register a single display, but OpenGL and Vulkan should both be working in FurMark 🤯 
- Never close FurMark, just keep it running in the background, not even running benchmarks, just open

If I ever close FurMark, things go back to broken, not sure why. I would assume that it is somehow keeping the graphics context up and available 🤔 

Anyway, now I can freely disconnect/connect and not have to worry about juggling multiple monitors or ensuring the the graphics APIs are available and all my apps seem to work just fine 🎉 Hope this helps for those running into similar issues.

**Comment by TimotheFCN** [comment]
> In trying out the above, I ran into some issues but found what seems to be an optimal solution for myself. I have no idea why this is working for me, but here are the tweaked steps that seem to work:
> 
> * Connect via Parsec / Moonlight
> * Install the extra display with [IddSampleDriver](https://github.com/ge9/IddSampleDriver) via scoop.
>   
>   * Not sure if this is only needed because I'm using Sunshine
> * Install [FurMark](https://www.geeks3d.com/furmark/) to help detect if OpenGL and Vulkan are available
>   
>   * Running it at this point should fail
> * Re-enable the `Microsoft Hyper-V Video` Display adapter in Device manager
> * Connect to the vm display via Hyper-V, while still connected via (Parsec / Moonlight)
>   
>   * Windows should now register two displays
> * FurMark should now be able to open and you can confirm OpenGL is working
>   
>   * Vulkan doesn't work at this point for me
> * Close the display for Hyper-V
> * Now windows should only  register a single display, but OpenGL and Vulkan should both be working in FurMark 🤯
> * Never close FurMark, just keep it running in the background, not even running benchmarks, just open
> 
> If I ever close FurMark, things go back to broken, not sure why. I would assume that it is somehow keeping the graphics context up and available 🤔
> 
> Anyway, now I can freely disconnect/connect and not have to worry about juggling multiple monitors or ensuring the the graphics APIs are available and all my apps seem to work just fine 🎉 Hope this helps for those running into similar issues.

Since this works, it might be possible to create a simple script that runs at startup which enables the hyper-v display, launches a sample blank app requiring vulkan/opengl (but not rendering anything), then turns off the display driver
I will be looking into that

**Comment by mikrofyr** [comment]
This seems to work. I do see some stuttering in vulkan applications though. 
Same applications run smoothly outside hyper-v.

Wondering if there is some performance hit related to the workaround, or if it is just in general a performance issue with the hyper-v VM.

Several other graphics applications runs very smooth in the hyper-v VM. (they are not using opengl/vulkan as far as I know)

**Comment by nenkoru** [comment]
Have you fixed/encountered a problem with a Sunshine when it takes 3-10 times to actually connect without having an error about no video stream incoming or just pitch black screen?

**Comment by rooftopdandelion** [comment]
Hi can anyone post their windows version while getting this working?  I'm running into the same issue as @ichinoboy .  I've worked through different possible solutions (including @djblue 's) but the hyper-v display never becomes active.  Thanks in advance!

**Comment by csitangelo** [comment]
Thanks for this, Vulkan worked straight away after re-enabling Hyper-V Display Driver. Sunshine selects idd automatically :)

Fixed Yuzu, Stardew Valley, DOOM Eternal and any other game that couldn't detect a GPU

**Comment by timminator** [comment]
Do any of you who have managed to get it working have any more insights?  I'm facing the same problem as @waltmannz . Opengl works right away after enabling the Hyper-V-Monitor, but i am not able to get vulkan working, even after testing many variations. The tips mentioned by @djblue also didn't help.
A question at @simonlsp:  Where you using a specific Windows version on you host or in the virtual machine? I've noticed you alco commentated in this [thread](https://github.com/jamesstringerparsec/Easy-GPU-PV/issues/299), where a Windows Insider Build was mentioned for the VM.

**Comment by timminator** [comment]
I were able to get quite a bit further over the last week. I did quite a lot of testing with different Nvidia drivers as im using a GTX 1660 TI.
First of all i noticed that Nvidia seems to have added Vulkan support via GPU Paravirtualization in driver version 545.84, atleast for consumer graphics cards. The vkcube demo included in the VulkanSDK runs starting with this driver version. Since driver version 551.23 you can also successfully run vulkaninfo inside the VM. This indicates that Vulkan applications should be working in theory.
Now it gets interesting: I noticed that the included demos in GPU Caps Viewer run aswell as shown in the first post by @simonlsp. But the Geexlab demos in standalone were crashing. This seemed odd to me. The only difference between them were there architecture as GPU Caps Viewer is a 32-bit application. So i tried running the Furmark 32-bit version instead - and it ran perfectly fine:

![Screenshot 2024-08-09 012923 - Copy](https://github.com/user-attachments/assets/2cef23f8-542a-4fbb-9a62-b42b2c9dfac9)

Whats weird is that the vkcube demo however is a 64 bit application.
The 64 bit version of Furmark reports as a error message: "VK_ERROR_INITIALIZATION_FAILED (-3), unable to create the swapchain for GPU 0". I tried to investigate this further and created my own little Vulkan script that created a swapchain, but it worked compiled as a 32-bit and 64-bit application aswell. Maybe the issue is related to a certain Vulkan extension that is required by Furmark but not by the vkcube demo?
But it looks a lot like a driver issue.
I've also noticed another driver problem. The Geexlab demos and Furmark 32-bit only ran on a indirect display via virtual display driver or Parsec with Hyper-V display still enabled. Starting these two programs on the Hyper-V display resulted in a error message "VK_ERROR_DEVICE_LOST (-4)."
In short: It looks like only 32-bit Vulkan applications, which are more complex than just a spinning cube, work at the moment, .


Below is a brief summary of my driver tests (I've always did a clean driver install via DDU on the host and created a fresh VM for consistent results):

- 537.58: No Vulkan support, no Vulkan extensions in GPU Caps Viewer, vulkaninfo throws error

- 545.84 (October 17th, 2023): Vkcube demo runs, GPU Caps Viewer lists Vulkan extensions, GPU Caps Viewer demo and Furmark 32-bit run on second display, vulkaninfo throws error

- 551.23: Vulkaninfo can be executed in terminal

- 555.85: GPU is registered in GPU Caps Viewer instead of just an empty window

- 560.70: Didn't notice any difference

**Comment by UnstructuredFF** [comment]
> In trying out the above, I ran into some issues but found what seems to be an optimal solution for myself. I have no idea why this is working for me, but here are the tweaked steps that seem to work:
> 
> * Connect via Parsec / Moonlight
> * Install the extra display with [IddSampleDriver](https://github.com/ge9/IddSampleDriver) via scoop.
>   
>   * Not sure if this is only needed because I'm using Sunshine
> * Install [FurMark](https://www.geeks3d.com/furmark/) to help detect if OpenGL and Vulkan are available
>   
>   * Running it at this point should fail
> * Re-enable the `Microsoft Hyper-V Video` Display adapter in Device manager
> * Connect to the vm display via Hyper-V, while still connected via (Parsec / Moonlight)
>   
>   * Windows should now register two displays
> * FurMark should now be able to open and you can confirm OpenGL is working
>   
>   * Vulkan doesn't work at this point for me
> * Close the display for Hyper-V
> * Now windows should only  register a single display, but OpenGL and Vulkan should both be working in FurMark 🤯
> * Never close FurMark, just keep it running in the background, not even running benchmarks, just open
> 
> If I ever close FurMark, things go back to broken, not sure why. I would assume that it is somehow keeping the graphics context up and available 🤔
> 
> Anyway, now I can freely disconnect/connect and not have to worry about juggling multiple monitors or ensuring the the graphics APIs are available and all my apps seem to work just fine 🎉 Hope this helps for those running into similar issues.

I thank you greatly for this tutorial, I have Minecraft running in a VM with a smooth FPS of 300-500 
![image_2024-08-21_011700842](https://github.com/user-attachments/assets/c2888ec0-b889-4db7-9382-486ee8878d4b)

**Comment by timminator** [comment]
I want to update my findings once more. I've tested quite a few vulkan applications with interesting results. Some seem to be crashing like Furmark 64 bit, whereas other application run just fine. These are my results on a GTX 1660 TI:

**Runs:**
Detroit Become Human
Wolfenstein II: The New Colossus
Ghost Recon Breakpoint
Counter Strike 2
Dota 2
Geekbench 6
Furmark 32 bit

**Doesnt run:**
Quake 2 RTX
Doom 2016
Doom Eternal
The Crew Motorfest
Rainbox Six Extraction
GravityMark GPU Benchmark
3DMark
Furmark 64 bit

Games that support more than one API where explicitly tested with the Vulkan API. 
This does not help narrowing it down, but it reflects the current state.
It is interesting that Doom Eternal did not run on my setup, whereas @csitangelo mentioned that he was able to get it running. I am still really interested about the GPU you are using :-)

**Comment by nenkoru** [comment]
> I were able to get quite a bit further over the last week. I did quite a lot of testing with different Nvidia drivers as im using a GTX 1660 TI. First of all i noticed that Nvidia seems to have added Vulkan support via GPU Paravirtualization in driver version 545.84, atleast for consumer graphics cards. The vkcube demo included in the VulkanSDK runs starting with this driver version. Since driver version 551.23 you can also successfully run vulkaninfo inside the VM. This indicates that Vulkan applications should be working in theory. Now it gets interesting: I noticed that the included demos in GPU Caps Viewer run aswell as shown in the first post by @simonlsp. But the Geexlab demos in standalone were crashing. This seemed odd to me. The only difference between them were there architecture as GPU Caps Viewer is a 32-bit application. So i tried running the Furmark 32-bit version instead - and it ran perfectly fine:
> 
> ![Screenshot 2024-08-09 012923 - Copy](https://private-user-images.githubusercontent.com/150205162/357231645-2cef23f8-542a-4fbb-9a62-b42b2c9dfac9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU1Nzc4MzYsIm5iZiI6MTcyNTU3NzUzNiwicGF0aCI6Ii8xNTAyMDUxNjIvMzU3MjMxNjQ1LTJjZWYyM2Y4LTU0MmEtNGZiYi05YTYyLWI0MmIyYzlkZmFjOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTA1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkwNVQyMzA1MzZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02MGM5ZmFmNzIwYTE4ODcwY2ExYjFlY2ZmY2M0M2JiZjc0N2E3ODQ4MGNmNjYzNWMzMGNkZTY2NDkyNDA2NTk5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.SgCItCE8xd9V9ibxRNiMfwUH9IMwKUaduaBsiTU9bYs)
> 
> Whats weird is that the vkcube demo however is a 64 bit application. The 64 bit version of Furmark reports as a error message: "VK_ERROR_INITIALIZATION_FAILED (-3), unable to create the swapchain for GPU 0". I tried to investigate this further and created my own little Vulkan script that created a swapchain, but it worked compiled as a 32-bit and 64-bit application aswell. Maybe the issue is related to a certain Vulkan extension that is required by Furmark but not by the vkcube demo? But it looks a lot like a driver issue. I've also noticed another driver problem. The Geexlab demos and Furmark 32-bit only ran on a indirect display via virtual display driver or Parsec with Hyper-V display still enabled. Starting these two programs on the Hyper-V display resulted in a error message "VK_ERROR_DEVICE_LOST (-4)." In short: It looks like only 32-bit Vulkan applications, which are more complex than just a spinning cube, work at the moment, .
> 
> Below is a brief summary of my driver tests (I've always did a clean driver install via DDU on the host and created a fresh VM for consistent results):
> 
> * 537.58: No Vulkan support, no Vulkan extensions in GPU Caps Viewer, vulkaninfo throws error
> * 545.84 (October 17th, 2023): Vkcube demo runs, GPU Caps Viewer lists Vulkan extensions, GPU Caps Viewer demo and Furmark 32-bit run on second display, vulkaninfo throws error
> * 551.23: Vulkaninfo can be executed in terminal
> * 555.85: GPU is registered in GPU Caps Viewer instead of just an empty window
> * 560.70: Didn't notice any difference

Reproduced the same behaviour. Furmark x86 Vulkan works fine, but the x64. 

Seems more like a driver issue according to different issues on github concerning the same error log about a swapchain .
I am pretty much sure then those games executables that you mentioned are x86. And the others are x64.
OpenGL works fine tho on both.

Host: Windows Server 2022
Guest: Windows 10 22H2

**Comment by nenkoru** [comment]
It's worth noting that this whole thing with the pass through works because of WDDM and Microsoft forcing vendors to create drivers comply with it.
Here is a good comment on Reddit.

https://www.reddit.com/r/VFIO/comments/1300rlk/comment/k6wvkd1

**Comment by Am-ilcar** [comment]
> In trying out the above, I ran into some issues but found what seems to be an optimal solution for myself. I have no idea why this is working for me, but here are the tweaked steps that seem to work:
> 
> * Connect via Parsec / Moonlight
> * Install the extra display with [IddSampleDriver](https://github.com/ge9/IddSampleDriver) via scoop.
>   
>   * Not sure if this is only needed because I'm using Sunshine
> * Install [FurMark](https://www.geeks3d.com/furmark/) to help detect if OpenGL and Vulkan are available
>   
>   * Running it at this point should fail
> * Re-enable the `Microsoft Hyper-V Video` Display adapter in Device manager
> * Connect to the vm display via Hyper-V, while still connected via (Parsec / Moonlight)
>   
>   * Windows should now register two displays
> * FurMark should now be able to open and you can confirm OpenGL is working
>   
>   * Vulkan doesn't work at this point for me
> * Close the display for Hyper-V
> * Now windows should only  register a single display, but OpenGL and Vulkan should both be working in FurMark 🤯
> * Never close FurMark, just keep it running in the background, not even running benchmarks, just open
> 
> If I ever close FurMark, things go back to broken, not sure why. I would assume that it is somehow keeping the graphics context up and available 🤔
> 
> Anyway, now I can freely disconnect/connect and not have to worry about juggling multiple monitors or ensuring the the graphics APIs are available and all my apps seem to work just fine 🎉 Hope this helps for those running into similar issues.

wow! this worked for me! using parsec, but I had to first disconnect from Parsec, connect via HyperV, open FurMark and then close the HyperV window and connect again with Parsec!
![image](https://github.com/user-attachments/assets/26e0334f-e95c-4482-bc4a-7109dead66fc)

**Comment by Mancitiss** [comment]
I struggled to make FurMark works. After some hours I noticed that when connect using Hyper-V my screen was so small, so I think enhanced mode for Hyper-V was disabled somehow (it was enabled when it first created), so I turn on [Enhanced Session Mode for Hyper-V](https://learn.microsoft.com/en-us/windows-server/virtualization/hyper-v/learn-more/use-local-resources-on-hyper-v-virtual-machine-with-vmconnect) and then connect again and it prompts me to choose a window size instead of the small window I had, and after that FurMark (32bit) runs successfully

**Comment by Tobi-La** [comment]
> Since this works, it might be possible to create a simple script that runs at startup which enables the hyper-v display, launches a sample blank app requiring vulkan/opengl (but not rendering anything), then turns off the display driver
> I will be looking into that

@TimotheFCN Are there any updates for that?

**Comment by UnstructuredK** [comment]
> In trying out the above, I ran into some issues but found what seems to be an optimal solution for myself. I have no idea why this is working for me, but here are the tweaked steps that seem to work:
> 
> * Connect via Parsec / Moonlight
> * Install the extra display with [IddSampleDriver](https://github.com/ge9/IddSampleDriver) via scoop.
>   
>   * Not sure if this is only needed because I'm using Sunshine
> * Install [FurMark](https://www.geeks3d.com/furmark/) to help detect if OpenGL and Vulkan are available
>   
>   * Running it at this point should fail
> * Re-enable the `Microsoft Hyper-V Video` Display adapter in Device manager
> * Connect to the vm display via Hyper-V, while still connected via (Parsec / Moonlight)
>   
>   * Windows should now register two displays
> * FurMark should now be able to open and you can confirm OpenGL is working
>   
>   * Vulkan doesn't work at this point for me
> * Close the display for Hyper-V
> * Now windows should only  register a single display, but OpenGL and Vulkan should both be working in FurMark 🤯
> * Never close FurMark, just keep it running in the background, not even running benchmarks, just open
> 
> If I ever close FurMark, things go back to broken, not sure why. I would assume that it is somehow keeping the graphics context up and available 🤔
> 
> Anyway, now I can freely disconnect/connect and not have to worry about juggling multiple monitors or ensuring the the graphics APIs are available and all my apps seem to work just fine 🎉 Hope this helps for those running into similar issues.

Hey, I'm not sure if this'll help anyone although I hope it does, I've come back to this post

I made a very lightweight program that can run on startup to keep an instance of OpenGL alive when connected through Parsec or Moonlight with a video tutorial seriously hope this helps 

https://github.com/UnstructuredK/OpenGL-Keep-Alive

https://www.youtube.com/watch?v=aO2ujJGc59s

**Comment by Loyola-London** [comment]
> I'm trying to use this method but my Hyper-V console will freeze at the Windows loading screen and no matter what I do Windows will report the Hyper-V monitor as off and I cannot extend to it or enable it in any way.
> 
> Anyone know a fix for this? I'm using Apollo instead of Sunshine or Parsec and the monitor worked completely fine until I restarted and now I cannot get this trick to work.

I believe this method is no longer available as of Windows Cumulative Update 06-2025. The Hyper-V monitor in conjunction with GPU-P causes the VM to freeze on boot. If you have a virtual monitor installed...which you do...the VM will boot normally, but all attempts to display the HyperVMonitor will fail. 

See here: https://github.com/jamesstringerparsec/Easy-GPU-PV/issues/444

**Comment by bcw222** [comment]
> > I'm trying to use this method but my Hyper-V console will freeze at the Windows loading screen and no matter what I do Windows will report the Hyper-V monitor as off and I cannot extend to it or enable it in any way.
> > Anyone know a fix for this? I'm using Apollo instead of Sunshine or Parsec and the monitor worked completely fine until I restarted and now I cannot get this trick to work.
> 
> I believe this method is no longer available as of Windows Cumulative Update 06-2025. The Hyper-V monitor in conjunction with GPU-P causes the VM to freeze on boot. If you have a virtual monitor installed...which you do...the VM will boot normally, but all attempts to display the HyperVMonitor will fail.
> 
> See here: [#444](https://github.com/jamesstringer90/Easy-GPU-PV/issues/444)

The issue should be already solved.
